sw/device/tests/kmac_interrupt_test.c

```diff
@@ -0,0 +1,455 @@
+                   0x88c112b6, 0xf6dc4a95, 0xe1cfeffa, 0xa7809e6f, 0x3a796dcd,
+                   0xb5962e44, 0x179d6ff0, 0xc898c5a9, 0xd3f02195, 0x43623028,
+                   0x4c3a4fe7, 0x2fab7bda, 0x04e5b4d4, 0xe0420692, 0x32fcaa2a,
+                   0x05e92f07, 0xba0564ea, 0x7b169778, 0x61d4ca3e, 0x4a5d92ec,
+                   0x079cb3ba, 0x9a784e40, 0x6381498c, 0xed6d8b6a, 0x2be74d42,
+                   0xa234a3db, 0x60d10de8, 0xf0c77dda, 0xc8f94b72, 0x239a2bdf,
+                   0xbfeba4a6, 0xc91042e9, 0xa5a11310, 0x8b44d66a, 0xea9bff2f,
+                   0x441a445f, 0xe88ee35d, 0x89386c12, 0x1a8de11e, 0x46aff650,
+                   0x423323c9, 0xba7b8db4, 0x06c36eb0, 0x4fd75b36, 0xf0c70001,
+                   0x0aefb1df, 0x6ae399e6, 0xf71930a6, 0xdef2206,  0x5ce2a640,
+                   0x6a82fcf4, 0xa91b0815},
+        .digest_len = 102,  // Rate (r) is 34 words.
+    },
+};
+
+static dif_kmac_t kmac;
+static dif_rv_plic_t plic0;
+static const uint32_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+static volatile bool fifo_empty_irq = false;
+static volatile bool kmac_done_irq = false;
+
+/**
+ * KMAC interrupt handler
+ *
+ * Services KMAC interrupts, sets the appropriate flags that are used to
+ * determine success or failure of the test.
+ */
+static void handle_kmac_isr(const dif_rv_plic_irq_id_t interrupt_id) {
+  // NOTE: This initialization is superfluous, since the `default` case below
+  // is effectively noreturn, but the compiler is unable to prove this.
+  dif_kmac_irq_t kmac_irq = 0;
+
+  switch (interrupt_id) {
+    case kTopEarlgreyPlicIrqIdKmacFifoEmpty:
+      kmac_irq = kDifKmacIrqFifoEmpty;
+      LOG_INFO("FifoEmpty");
+      fifo_empty_irq = true;
+      break;
+    case kTopEarlgreyPlicIrqIdKmacKmacDone:
+      LOG_INFO("KmacDone");
+      kmac_irq = kDifKmacIrqKmacDone;
+      kmac_done_irq = true;
+      break;
+    case kTopEarlgreyPlicIrqIdKmacKmacErr:
+      kmac_irq = kDifKmacIrqKmacErr;
+      LOG_FATAL("KmacError!");
+      test_status_set(kTestStatusFailed);
+      break;
+    default:
+      LOG_FATAL("ISR is not implemented!");
+      test_status_set(kTestStatusFailed);
+  }
+
+  CHECK_DIF_OK(dif_kmac_irq_acknowledge(&kmac, kmac_irq));
+}
+
+/**
+ * External interrupt handler
+ *
+ * Handles all peripheral interrupts on Ibex. PLIC asserts an external interrupt
+ * line to the CPU, which results in a call to this handler. This handler
+ * overrides the default implementation, and prototype for this handler must
+ * include appropriate attributes.
+ */
+void handler_irq_external(void) {
+  // Claim the IRQ by reading the Ibex specific CC register.
+  dif_rv_plic_irq_id_t interrupt_id = 0;
+
+  CHECK_DIF_OK(dif_rv_plic_irq_claim(&plic0, kPlicTarget, &interrupt_id));
+
+  // Check if the interrupted peripheral is KMAC.
+  top_earlgrey_plic_peripheral_t peripheral_id =
+      top_earlgrey_plic_interrupt_for_peripheral[interrupt_id];
+  CHECK(peripheral_id == kTopEarlgreyPlicPeripheralKmac,
```

Comment:
Probably not an issue at the moment, but with OTTF we will be running tests in parallel (or at least will try to run as many as possible in parallel), as far as I understand. In which case we could potentially pick up some other interrupt here, in which case we probably just want skip servicing it instead of having a fatal failure?

I am might be overthinking it. @timothytrippel could you please sanity check the above? Thanks!

Reply:
The purpose of the OTTF is **_not_** to run several tests in parallel. The purpose of the OTTF is to provide support at the test framework level for both bare-metal tests (which are all chip-level tests documented in the test plan [here](https://docs.opentitan.org/hw/top_earlgrey/doc/dv/), and (future) system level scenario tests (which are currently in the planning phase) that will have some concurrency tests (i.e. tests which launch several execution contexts, i.e., "tasks", on FreeRTOS. So for the purpose of non-concurrent tests this is perfectly ok, as this is a bare-metal test.

Reply:
Does that make sense?